### ISSUE TITLE
Investigate Observables

### ISSUE THREAD

**Issue by imhoffd** [driftyco/ionic-cloud-angular#4]
As suggested in https://github.com/driftyco/ionic-cloud-angular/issues/1#issuecomment-229682787, and similar to https://github.com/driftyco/ionic-cloud/issues/4, the cloud client should somehow provide observable streams to events such as "is the user logged in" and "push notification received." 

Investigate how this would be done and discuss options.

@cklanac @tlancina

**Comment by justinschuldt** [comment]
Here is our use case for push notification observables.

We have a couple different cases for what to do when a notification is received:
New message for the user - go to the message details page and load the message
New update available (via Deploy) - update and then close the app
Silent notification - do some background stuff, download new stuff, etc

I'd like to have multiple subscribers to the stream so that the business logic for a feature can be contained in its component(s). For example all the logic to determine if an update should be downloaded and extracted should be contained in the DeployService, all the logic to navigate to and show a new message should be in the MessagePage class.

I'd like to avoid having a bunch of business logic and dependencies where the Push.init happens like:

``` ts
push.on('notification', data => {
  if (data.additionalData.messageId) {
    nav.push(MessagePage, {id: data.additionalData.messageId});
  } else if (data.additionalData.deploy === true) {
    DeployService.update();
  } else if (newFeature) (
    // bunch of logic
  } else {
    // do some background stuff
  }
});
```

Right now I've written a class (PushService) that inits the push plugin and returns observables that I consume like this:

``` ts
import {NavController} from 'ionic-angular';
import {Component} from '@angular/core';

import {AzureService} from '../../shared';
import {MessagePage} from '../../dashboard';
import {PushService} from './';
import {Observable} from 'rxjs/Observable';

@Component({
    templateUrl: 'build/shared/native/native.html',
    providers: [PushService]
})
export class NativePage {
    constructor(
        private nav: NavController,
        private azureService: AzureService,
        private pushService: PushService
    ) {

        this.pushService.notification()
            .subscribe(
                data => {
                    console.debug('subscribed data: ', data);
                    azureService.api('pushRecieved').post(data);
                    if (data && data.additionalData && data.additionalData.messagesId)
                        nav.push(MessagePage, {id: data.additionalData.messagesId});
                }
            );
    }

    // Push
    register(): void {
        this.pushService.register();
    }
}
```

Right now my PushService is doing the observable wrapping, but I think that wrapping should happen at a lower level. Maybe Ionic Native should included the observable wrapper? 

tl;dr Observables would be cool, but I think it should be done at the plugin level for push notifications. 

Let me know what you think!

**Comment by imhoffd** [comment]
@justinschuldt Can I have you update and test observable support for push notifications? (FYI there are breaking changes.) I only did it for push notifications because I'd like you to validate my approach.

``` typescript
import { Push } from '@ionic/cloud-angular';
```

``` typescript
this.push.rx.notification()
  .subscribe((msg) => {
    console.log('OBSERVABLE got notification!', msg);
  });
```

I was able to use `filter()` on a deploy flag to simulate one of your use cases.

Check it out: https://github.com/driftyco/ionic-cloud-angular/blob/master/src/index.ts#L21-L38

You may notice there's no support for dispose yet. I'm waiting on an issue whose URL I left in the comment.

Likes/dislikes?

**Comment by imhoffd** [comment]
What other observables might be suited in the Cloud Client? I know it would be nice to subscribe to a deploy channel, but the deploy plugin doesn't let us _push_ out changes (we'd have to poll).

**Comment by justinschuldt** [comment]
Hey, sorry it's taken me a while to get to this.

Reading through it, it looks great. I'm going to spin up a new app with the cloud client and test it out this weekend.

I've been testing my push-plugin wrapper service in different app states, and have gotten some inconsistent behavior when the app is dead. My criteria has been that when the phone receives a notification, it is able to send an http request. 

I'll test yours in the same conditions and let you know what I find!

**Comment by justinschuldt** [comment]
As far as observables for other parts of the cloud client, I'm not sure an observable deploy service would be very valuable to most people. I cant think of a realistic situation where I would want to watch for updates that vigilantly.

it could be cool to see analytics in real time, but that wouldn't be needed in the client.
